File: backend/src/ssm/api/v2/diffs.py
"""``/api/v2/diffs/{host_id}`` — compare the host's authorized_keys to the DB.

For each distinct ``login`` authorized on the host we:
  1. Build the set of key lines the DB says should be present for that login.
  2. Read ``/home/<login>/.ssh/authorized_keys`` over SSH.
  3. Return one :class:`KeyDiff` per expected-or-observed line, tagged as
     ``present``, ``missing_on_host``, or ``extra_on_host``.

Disabled hosts short-circuit with ``HOST_DISABLED`` before any SSH happens.
"""

from __future__ import annotations

from enum import StrEnum
from typing import Annotated

from fastapi import Depends
from pydantic import BaseModel
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from ssm.auth.deps import protected_router
from ssm.core.envelope import ApiResponse
from ssm.core.errors import HostNotFound, SshConnectFailed
from ssm.db.deps import db_session
from ssm.db.models import Authorization, Host, UserKey
from ssm.ssh.deps import get_ssh_client
from ssm.ssh.protocol import SshClient, SshTarget
from ssm.ssh.safety import ensure_host_not_disabled, ensure_writable

router = protected_router(prefix="/diffs", tags=["diffs"])


class DiffStatus(StrEnum):
    PRESENT = "present"
    MISSING_ON_HOST = "missing_on_host"
    EXTRA_ON_HOST = "extra_on_host"


class KeyDiff(BaseModel):
    status: DiffStatus
    line: str


class LoginDiff(BaseModel):
    login: str
    read_error: str | None = None
    items: list[KeyDiff]


class HostDiff(BaseModel):
    host_id: int
    host_name: str
    disabled: bool
    logins: list[LoginDiff]


def _format_key_line(key_type: str, key_base64: str, label: str | None) -> str:
    if label:
        return f"{key_type} {key_base64} {label}"
    return f"{key_type} {key_base64}"


def _parse_authorized_keys(text: str) -> list[str]:
    """Strip comments and blank lines; preserve the remaining verbatim."""
    lines: list[str] = []
    for raw in text.splitlines():
        stripped = raw.strip()
        if not stripped or stripped.startswith("#"):
            continue
        lines.append(stripped)
    return lines


async def _build_target(session: AsyncSession, host: Host) -> SshTarget:
    jump: SshTarget | None = None
    if host.jump_via is not None:
        jump_host = await session.get(Host, host.jump_via)
        if jump_host is not None:
            jump = await _build_target(session, jump_host)
    return SshTarget(
        host_id=host.id,
        name=host.name,
        address=host.address,
        port=host.port,
        username=host.username,
        jump_target=jump,
    )


async def _expected_keys_for_login(
    session: AsyncSession, host_id: int, login: str
) -> tuple[list[int], list[str]]:
    """Return (user_ids authorized as ``login``, expected key lines)."""
    stmt = select(Authorization.user_id).where(
        Authorization.host_id == host_id, Authorization.login == login
    )
    user_ids = list((await session.execute(stmt)).scalars().all())
    if not user_ids:
        return [], []

    keys_stmt = select(UserKey).where(UserKey.user_id.in_(user_ids))
    rows = (await session.execute(keys_stmt)).scalars().all()
    lines = [_format_key_line(k.key_type, k.key_base64, k.name) for k in rows]
    return user_ids, lines


async def _diff_for_login(
    client: SshClient, target: SshTarget, login: str, expected: list[str]
) -> LoginDiff:
    expected_set = set(expected)
    actual_set: set[str] = set()
    read_error: str | None = None

    path = f"/home/{login}/.ssh/authorized_keys"
    try:
        file = await client.read_file(target, path)
        actual_set = set(_parse_authorized_keys(file.content))
    except SshConnectFailed as exc:
        read_error = str(exc)

    items: list[KeyDiff] = []
    for line in sorted(expected_set & actual_set):
        items.append(KeyDiff(status=DiffStatus.PRESENT, line=line))
    for line in sorted(expected_set - actual_set):
        items.append(KeyDiff(status=DiffStatus.MISSING_ON_HOST, line=line))
    for line in sorted(actual_set - expected_set):
        items.append(KeyDiff(status=DiffStatus.EXTRA_ON_HOST, line=line))
    return LoginDiff(login=login, read_error=read_error, items=items)


@router.get("/{host_id}", response_model=ApiResponse[HostDiff])
async def get_host_diff(
    host_id: int,
    session: Annotated[AsyncSession, Depends(db_session)],
    ssh: Annotated[SshClient, Depends(get_ssh_client)],
) -> ApiResponse[HostDiff]:
    host = await session.get(Host, host_id)
    if host is None:
        raise HostNotFound(f"host {host_id} not found")
    ensure_host_not_disabled(disabled=host.disabled, host_name=host.name)

    target = await _build_target(session, host)

    logins_stmt = (
        select(Authorization.login)
        .where(Authorization.host_id == host_id)
        .distinct()
        .order_by(Authorization.login)
    )
    logins = list((await session.execute(logins_stmt)).scalars().all())

    login_diffs: list[LoginDiff] = []
    for login in logins:
        _, expected = await _expected_keys_for_login(session, host_id, login)
        login_diffs.append(await _diff_for_login(ssh, target, login, expected))

    return ApiResponse[HostDiff].ok(
        HostDiff(host_id=host.id, host_name=host.name, disabled=host.disabled, logins=login_diffs)
    )


class SyncedLogin(BaseModel):
    login: str
    written_keys: int


class SyncResult(BaseModel):
    host_id: int
    host_name: str
    logins: list[SyncedLogin]


@router.post("/{host_id}/sync", response_model=ApiResponse[SyncResult])
async def sync_host(
    host_id: int,
    session: Annotated[AsyncSession, Depends(db_session)],
    ssh: Annotated[SshClient, Depends(get_ssh_client)],
) -> ApiResponse[SyncResult]:
    host = await session.get(Host, host_id)
    if host is None:
        raise HostNotFound(f"host {host_id} not found")
    ensure_host_not_disabled(disabled=host.disabled, host_name=host.name)

    target = await _build_target(session, host)

    logins_stmt = (
        select(Authorization.login)
        .where(Authorization.host_id == host_id)
        .distinct()
        .order_by(Authorization.login)
    )
    logins = list((await session.execute(logins_stmt)).scalars().all())

    # Fail loudly and early on any readonly login so we never partial-write.
    for login in logins:
        await ensure_writable(ssh, target, login)

    synced: list[SyncedLogin] = []
    for login in logins:
        _, expected = await _expected_keys_for_login(session, host_id, login)
        content = "".join(f"{line}\n" for line in expected)
        path = f"/home/{login}/.ssh/authorized_keys"
        await ssh.write_file(target, path, content)
        synced.append(SyncedLogin(login=login, written_keys=len(expected)))

    return ApiResponse[SyncResult].ok(
        SyncResult(host_id=host.id, host_name=host.name, logins=synced)
    )

File: phoenix/lib/ssm/diffs.ex
defmodule Ssm.Diffs do
  @moduledoc """
  Compare a host's on-disk `authorized_keys` to what the database authorizes,
  and push the DB state to the host — port of api/v2/diffs.py.

  Every read/write goes through `Ssm.Ssh.ScriptRunner` (never raw SFTP), so
  home-dir probing, `has_pragma` detection, the readonly sentinel and managed
  backups all come for free. Disabled hosts short-circuit before any SSH.
  """

  import Ecto.Query

  alias Ssm.Authorizations.Authorization
  alias Ssm.Hosts
  alias Ssm.Hosts.Host
  alias Ssm.Repo
  alias Ssm.Ssh.ScriptRunner
  alias Ssm.Users.UserKey

  defmodule KeyDiff do
    @moduledoc "One key line and its presence status."
    @enforce_keys [:status, :line]
    defstruct [:status, :line]
    @type status :: :present | :missing_on_host | :extra_on_host
    @type t :: %__MODULE__{status: status(), line: String.t()}
  end

  defmodule LoginDiff do
    @moduledoc "Per-login diff of authorized_keys."
    defstruct login: nil,
              has_pragma: false,
              readonly_condition: nil,
              read_error: nil,
              items: []
  end

  defmodule HostDiff do
    @moduledoc "A host's full diff across every login."
    defstruct [:host_id, :host_name, :disabled, logins: []]
  end

  @doc """
  Compute the diff for a host. Returns `{:error, {:host_disabled, _}}` for a
  disabled host and `{:error, :not_found}` for an unknown one.
  """
  @spec host_diff(module(), integer()) ::
          {:ok, HostDiff.t()} | {:error, term()}
  def host_diff(client \\ Ssm.Ssh, host_id) do
    with {:ok, host} <- Hosts.fetch_host(host_id),
         :ok <- ensure_not_disabled(host) do
      target = Hosts.target_for(host)

      {observed, read_error} =
        case ScriptRunner.get_ssh_keyfiles(client, target) do
          {:ok, entries} -> {Map.new(entries, &{&1.login, &1}), nil}
          {:error, {_code, message}} -> {%{}, message}
        end

      expected_logins = expected_logins(host_id)
      all_logins = (expected_logins ++ Map.keys(observed)) |> Enum.uniq() |> Enum.sort()

      logins =
        Enum.map(all_logins, fn login ->
          build_login_diff(host_id, login, Map.get(observed, login), read_error)
        end)

      {:ok,
       %HostDiff{host_id: host.id, host_name: host.name, disabled: host.disabled, logins: logins}}
    end
  end

  defmodule SyncedLogin do
    @moduledoc false
    defstruct [:login, :written_keys]
  end

  @doc """
  Push the DB's expected `authorized_keys` to every authorized login on the
  host. Bails atomically on the first readonly login (the script refuses the
  write) — never partial-writes. Disabled host short-circuits.
  """
  @spec sync_host(module(), integer()) ::
          {:ok, [SyncedLogin.t()]} | {:error, term()}
  def sync_host(client \\ Ssm.Ssh, host_id) do
    with {:ok, host} <- Hosts.fetch_host(host_id),
         :ok <- ensure_not_disabled(host) do
      target = Hosts.target_for(host)

      host_id
      |> expected_logins()
      |> Enum.sort()
      |> sync_logins(client, target, host_id, [])
    end
  end

  defp sync_logins([], _client, _target, _host_id, acc), do: {:ok, Enum.reverse(acc)}

  defp sync_logins([login | rest], client, target, host_id, acc) do
    expected = expected_keys_for_login(host_id, login)
    content = Enum.map_join(expected, "", &(&1 <> "\n"))

    case ScriptRunner.set_authorized_keyfile(client, target, login, content) do
      :ok ->
        synced = %SyncedLogin{login: login, written_keys: length(expected)}
        sync_logins(rest, client, target, host_id, [synced | acc])

      {:error, _} = error ->
        error
    end
  end

  ## Diff building

  defp build_login_diff(host_id, login, entry, read_error) do
    expected = expected_keys_for_login(host_id, login)
    actual = if entry, do: parse_authorized_keys(entry.keyfile), else: []

    %LoginDiff{
      login: login,
      has_pragma: !!(entry && entry.has_pragma),
      readonly_condition: entry && entry.readonly_condition,
      read_error: if(is_nil(entry) && read_error, do: read_error),
      items: compute_diff(expected, actual)
    }
  end

  @doc false
  def compute_diff(expected, actual) do
    exp = MapSet.new(expected)
    act = MapSet.new(actual)

    present = exp |> MapSet.intersection(act) |> to_sorted(:present)
    missing = exp |> MapSet.difference(act) |> to_sorted(:missing_on_host)
    extra = act |> MapSet.difference(exp) |> to_sorted(:extra_on_host)

    present ++ missing ++ extra
  end

  defp to_sorted(set, status) do
    set |> Enum.sort() |> Enum.map(&%KeyDiff{status: status, line: &1})
  end

  @doc false
  def format_key_line(key_type, key_base64, nil), do: "#{key_type} #{key_base64}"
  def format_key_line(key_type, key_base64, ""), do: "#{key_type} #{key_base64}"
  def format_key_line(key_type, key_base64, label), do: "#{key_type} #{key_base64} #{label}"

  defp parse_authorized_keys(text) do
    text
    |> String.split("\n")
    |> Enum.map(&String.trim/1)
    |> Enum.reject(&(&1 == "" or String.starts_with?(&1, "#")))
  end

  ## Expected-state queries

  defp expected_logins(host_id) do
    Repo.all(
      from a in Authorization,
        where: a.host_id == ^host_id,
        distinct: true,
        select: a.login
    )
  end

  @doc false
  def expected_keys_for_login(host_id, login) do
    user_ids =
      Repo.all(
        from a in Authorization,
          where: a.host_id == ^host_id and a.login == ^login,
          select: a.user_id
      )

    case user_ids do
      [] ->
        []

      ids ->
        Repo.all(
          from k in UserKey,
            where: k.user_id in ^ids,
            select: {k.key_type, k.key_base64, k.name}
        )
        |> Enum.map(fn {type, base64, name} -> format_key_line(type, base64, name) end)
    end
  end

  defp ensure_not_disabled(%Host{disabled: true, name: name}),
    do: {:error, {:host_disabled, "host #{inspect(name)} is disabled"}}

  defp ensure_not_disabled(%Host{}), do: :ok
end
